apps/web/src/components/EmailVerification/EmailVerification.tsx

```diff
@@ -0,0 +1,288 @@
+import { useState, useEffect, useReducer } from "react";
+import { useIntl } from "react-intl";
+import CheckBadgeIcon from "@heroicons/react/24/outline/CheckBadgeIcon";
+import { FormProvider, SubmitHandler, useForm } from "react-hook-form";
+import { useMutation } from "urql";
+
+import { Button, Heading } from "@gc-digital-talent/ui";
+import { Input, Submit } from "@gc-digital-talent/forms";
+import { errorMessages } from "@gc-digital-talent/i18n";
+import { toast } from "@gc-digital-talent/toast";
+import { graphql } from "@gc-digital-talent/graphql";
+import { useLogger } from "@gc-digital-talent/logger";
+import { useAuthorization } from "@gc-digital-talent/auth";
+
+const SendUserEmailVerification_Mutation = graphql(/* GraphQL */ `
+  mutation SendUserEmailVerification($id: ID!) {
+    sendUserEmailVerification(id: $id) {
+      id
+      emailVerifiedAt
+    }
+  }
+`);
+
+const VerifyUserEmail_Mutation = graphql(/* GraphQL */ `
+  mutation VerifyUserEmail($id: ID!, $code: String!) {
+    verifyUserEmail(id: $id, code: $code) {
+      id
+      emailVerifiedAt
+    }
+  }
+`);
+
+const getTitle = (
+  emailType: NonNullable<EmailVerificationProps["emailType"]>,
+) => {
+  switch (emailType) {
+    // presumably we'll have more than one type eventually
+    case "contact":
+    default:
+      return {
+        defaultMessage: "Verify your contact email",
+        id: "TguSOt",
+        description: "Heading for email verification form",
+      };
+  }
+};
+
+const CODE_REQUEST_THROTTLE_DELAY_MS = 1000 * 60;
+
+type FormValues = {
+  verificationCode: string;
+};
+
+export interface EmailVerificationProps {
+  emailType?: "contact";
+  // The email address that the code was sent to.  Displayed to the user.
+  emailAddress: string;
+  // Event if verification is successful.
+  onVerificationSuccess: () => void;
+  // Event if they choose to skip.  Skip button removed if prop not provided.
+  onSkip?: () => void;
+}
+
+const EmailVerification = ({
+  emailType = "contact",
+  emailAddress,
+  onVerificationSuccess,
+  onSkip,
+}: EmailVerificationProps) => {
+  const intl = useIntl();
+  const logger = useLogger();
+  const { userAuthInfo } = useAuthorization();
+  const [, executeSendEmailMutation] = useMutation(
+    SendUserEmailVerification_Mutation,
+  );
+  const [, executeVerifyUserEmailMutation] = useMutation(
+    VerifyUserEmail_Mutation,
+  );
+  const methods = useForm<FormValues>({});
+
+  const [noCodeRequestUntil, setNoCodeRequestUntil] = useState<number>(0);
+  const msUntilCanRequestACode = Math.max(0, noCodeRequestUntil - Date.now());
+  const canRequestACode = msUntilCanRequestACode === 0;
+  const [, forceUpdate] = useReducer((x) => x + 1, 0); // https://legacy.reactjs.org/docs/hooks-faq.html#is-there-something-like-forceupdate
```

Comment:
I know the react docs don't recommend this but I can't think of a better way.  The only alternative I can think of is to add a junk `useState` like "should rerender" and I find this solution more expressive.

Reply:
I could be wrong but, could we not just use a timeout? :thinking: 

```tsx
// Instead of
setNoCodeRequestUntil(...);

// Do this
setCanRequestCode(false);

useEffect(() => {
  let timerId;
  
  if(!canReuqestCode) {
    timerId = setTimeout(() => {
      setCanRequestCode(true);
     }, CODE_REQUEST_THROTTLE_DELAY_MS);
  }
  
  return () => clearTimeout(timerId);
}, [canRequestCode]);
```

We do something similar in the [polling query hook](https://github.com/GCTC-NTGC/gc-digital-talent/blob/7cce90c6616ba0977b984c47031ee81e4ea73da2/apps/web/src/hooks/usePollingQuery.ts#L15-L28).